builder/solver_monitor.go

```diff
@@ -264,6 +287,60 @@ func (sm *solverMonitor) printHeader(vm *vertexMonitor) {
 	vm.printHeader(!seen || sm.verbose)
 }
 
+func (sm *solverMonitor) recordTiming(targetStr, targetBrackets, salt string, vertex *client.Vertex) {
+	if vertex.Started == nil || vertex.Completed == nil {
+		return
+	}
+	dur := vertex.Completed.Sub(*vertex.Started)
+	if dur == 0 {
+		return
+	}
+	key := timingKey{
+		targetStr:      targetStr,
+		targetBrackets: targetBrackets,
+		salt:           salt,
+	}
+	sm.timingTable[key] = sm.timingTable[key] + dur
+}
+
+func (sm *solverMonitor) PrintTiming() {
+	if !sm.verbose {
+		return
+	}
+	sm.console.
+		WithMetadataMode(true).
+		Printf("Summary of timing information\n" +
+			"Note that the times do not include the expansion of commands like BUILD, FROM, COPY (artifact).")
+	var total time.Duration
+	durs := make([]time.Duration, 0, len(sm.timingTable))
+	durMap := make(map[time.Duration][]timingKey)
```

Comment:
I don't understand why the code is bucketing this by duration? is there a reason we can't just sort sm.timingTable directly (or make a copy of it, then sort it)? Then when it's printed to the console, one would only need to iterate over a single for loop rather than have two nested loops.

Reply:
Not really... maps in Go are not sorted. So we have to collect values in a slice and sort that slice and then reverse lookup. Don't know if you have something else in mind

Reply:
The double for loop is only for cases where multiple `timingKey`s have the exact same duration (very rare). It's not really inefficient.

Reply:
Ah, I didn't realise `sm.timingTable` was a map; I was wondering if we could do something like this instead? https://github.com/earthly/earthly/pull/533/files

I found the use of durMap hard to follow.

Reply:
full disclosure, I haven't tested that code; I'm also fine if you want to keep your code as-is.

Reply:
Yeah, much nicer. I would also move

```
type durationAndKey struct {
	dur time.Duration
	key timingKey
}
```

within the function itself since it's only used there.